samza-azure/src/test/java/org/apache/samza/system/azureblob/avro/TestAzureBlobOutputStream.java

```diff
@@ -0,0 +1,347 @@
+    when(mockBlobAsyncClient.stageBlock(anyString(), any(), anyLong())).thenReturn(
+        Mono.just(new SimpleResponse(null, 200, null, null)).flatMap(FluxUtil::toMono));
+    when(mockBlobAsyncClient.commitBlockListWithResponse(any(), any(), any(), any(), any())).thenReturn(
+        Mono.just(new SimpleResponse(null, 200, null, null)));
+
+    when(mockBlobAsyncClient.getBlobUrl()).thenReturn("https://samza.blob.core.windows.net/fake-blob-url");
+
+    mockMetrics = mock(AzureBlobWriterMetrics.class);
+
+    mockCompression = mock(Compression.class);
+    doReturn(compressedBytes).when(mockCompression).compress(bytes);
+
+    azureBlobOutputStream = spy(new AzureBlobOutputStream(mockBlobAsyncClient, threadPool, mockMetrics,
+        60000, THRESHOLD, mockByteArrayOutputStream, mockCompression));
+  }
+
+  @Test
+  public void testWrite() {
+    byte[] b = new byte[THRESHOLD - 10];
+    azureBlobOutputStream.write(b, 0, THRESHOLD - 10);
+    verify(mockBlobAsyncClient, never()).stageBlock(any(), any(), anyLong()); // since size of byte[] written is less than threshold
+    verify(mockMetrics).updateWriteByteMetrics(THRESHOLD - 10);
+    verify(mockMetrics, never()).updateAzureUploadMetrics();
+  }
+
+  @Test
+  public void testWriteLargerThanThreshold() {
+    byte[] largeRecord = randomString.substring(0, 2 * THRESHOLD).getBytes();
+    byte[] largeRecordFirstHalf = randomString.substring(0, THRESHOLD).getBytes();
+    byte[] largeRecordSecondHalf = randomString.substring(THRESHOLD, 2 * THRESHOLD).getBytes();
+
+    byte[] compressB = randomString.substring(0, THRESHOLD / 2).getBytes();
+
+    doReturn(compressB).when(mockCompression).compress(largeRecordFirstHalf);
+    doReturn(compressB).when(mockCompression).compress(largeRecordSecondHalf);
+
+    azureBlobOutputStream.write(largeRecord, 0, 2 * THRESHOLD);
+    // azureBlobOutputStream.close waits on the CompletableFuture which does the actual stageBlock in uploadBlockAsync
+    azureBlobOutputStream.close();
+
+    // invoked 2 times for the data which is 2*threshold
+    verify(mockCompression).compress(largeRecordFirstHalf);
+    verify(mockCompression).compress(largeRecordSecondHalf);
+    ArgumentCaptor<Flux> argument = ArgumentCaptor.forClass(Flux.class);
+    verify(mockBlobAsyncClient).stageBlock(eq(blockIdEncoded(0)), argument.capture(), eq((long) compressB.length));
+    verify(mockBlobAsyncClient).stageBlock(eq(blockIdEncoded(1)), argument.capture(), eq((long) compressB.length));
+    argument.getAllValues().forEach(flux -> {
+        Assert.assertEquals(ByteBuffer.wrap(compressB), flux.blockFirst());
+      });
+    verify(mockMetrics).updateWriteByteMetrics(2 * THRESHOLD);
+    verify(mockMetrics, times(2)).updateAzureUploadMetrics();
+  }
+
+  @Test
+  public void testWriteLargeRecordWithSmallRecordInBuffer() {
+    byte[] halfBlock = new byte[THRESHOLD / 2];
+    byte[] fullBlock = new byte[THRESHOLD];
```

Comment:
Would it be good to fill in these blocks with some data to be able to verify that the expected data gets written? The test currently just checks the size of the blocks through the byte array comparisons, but doesn't verify that the expected data is written as expected.

Reply:
`compressB = randomString.substring(0, THRESHOLD / 2).getBytes();
..
Assert.assertEquals(ByteBuffer.wrap(compressB), flux.blockFirst());`

data in compressB is compared to the argument passed to stageBlock. This does check the actual data passed to stageBlock. Hence it is checking that the expected data is written as expected.

Reply:
It looks like you use the bytes in `testWriteLargerThanThreshold`, but not in `testWriteLargeRecordWithSmallRecordInBuffer`. If you believe that `testWriteLargeRecordWithSmallRecordInBuffer` does not need to use bytes due to already using bytes in `testWriteLargerThanThreshold`, please add a comment to `testWriteLargeRecordWithSmallRecordInBuffer`.